File: my-project/mab-agent.cc
/*
 * MATLAB RA project migration: MAB implementation.
 */

#include "mab-agent.h" // 引入类声明。

#include <algorithm> // 使用 max。
#include <numeric> // 使用 accumulate。

namespace myproject // 实现项目命名空间。
{

MabAgent::MabAgent(uint32_t numActions, uint32_t windowSize, double epsilon) // 构造 MAB 智能体。
    : m_numActions(numActions), // 保存动作数量。
      m_windowSize(windowSize), // 保存滑动窗口长度。
      m_epsilon(epsilon) // 保存探索概率。
{
    for (double edge = -5.0; edge <= 35.0; edge += 5.0) // 构建 MATLAB 的 -5:5:35 边界。
    {
        m_snrEdges.push_back(edge); // 保存一个边界。
    }
    m_numBins = static_cast<uint32_t>(m_snrEdges.size()) + 1; // MATLAB NumBins=length(edges)+1。
    m_uniformRv = ns3::CreateObject<ns3::UniformRandomVariable>(); // 创建均匀随机变量。
    m_uniformRv->SetAttribute("Min", ns3::DoubleValue(0.0)); // 设置下界。
    m_uniformRv->SetAttribute("Max", ns3::DoubleValue(1.0)); // 设置上界。
    Reset(); // 初始化奖励缓冲区。
}

void MabAgent::Reset() // 重置 MAB 缓冲区。
{
    m_rewardBuffers.assign(m_numBins, std::vector<std::deque<double>>(m_numActions)); // 创建桶数乘动作数的空 deque 表。
}

uint32_t MabAgent::SelectAction(double snrDb) // 选择动作。
{
    const uint32_t bin = GetStateBin(snrDb); // 计算当前 SNR 桶。
    std::vector<uint32_t> untested; // 保存当前桶中尚未尝试的动作。
    for (uint32_t actionIndex = 1; actionIndex <= m_numActions; ++actionIndex) // 遍历所有动作。
    {
        if (m_rewardBuffers[bin][actionIndex - 1].empty()) // 如果该动作没有样本。
        {
            untested.push_back(actionIndex); // 记录为未测试动作。
        }
    }
    if (!untested.empty()) // 如果存在未测试动作。
    {
        const uint32_t pick = m_uniformRv->GetInteger(0, static_cast<uint32_t>(untested.size() - 1)); // 随机选择一个未测试动作。
        return untested[pick]; // 返回该动作。
    }
    if (m_uniformRv->GetValue() < m_epsilon) // 按 epsilon 判断是否探索。
    {
        return m_uniformRv->GetInteger(1, m_numActions); // 探索阶段随机选择动作。
    }
    double bestReward = -1.0; // 初始化最佳平均奖励。
    uint32_t bestAction = 1; // 初始化最佳动作。
    for (uint32_t actionIndex = 1; actionIndex <= m_numActions; ++actionIndex) // 遍历所有动作。
    {
        const double reward = MeanReward(bin, actionIndex); // 计算当前动作平均奖励。
        if (reward > bestReward) // 判断是否更优。
        {
            bestReward = reward; // 更新最佳奖励。
            bestAction = actionIndex; // 更新最佳动作。
        }
    }
    return bestAction; // 返回最佳动作。
}

void MabAgent::Update(double snrDb, uint32_t actionIndex, double reward) // 更新奖励窗口。
{
    NS_ABORT_MSG_IF(actionIndex == 0 || actionIndex > m_numActions, "Action index must be in [1, NumActions]"); // 防止动作越界。
    const uint32_t bin = GetStateBin(snrDb); // 计算 SNR 桶。
    std::deque<double>& buffer = m_rewardBuffers[bin][actionIndex - 1]; // 获取对应奖励缓冲区。
    buffer.push_back(reward); // 压入新奖励。
    if (buffer.size() > m_windowSize) // 如果超过滑动窗口长度。
    {
        buffer.pop_front(); // 删除最旧样本。
    }
}

uint32_t MabAgent::GetStateBin(double snrDb) const // 映射 SNR 桶。
{
    for (uint32_t i = 0; i < m_snrEdges.size(); ++i) // 遍历分桶边界。
    {
        if (snrDb <= m_snrEdges[i]) // MATLAB discretize 的边界近似处理。
        {
            return i; // 返回 0 基桶索引。
        }
    }
    return m_numBins - 1; // 高于最高边界时返回最后一桶。
}

double MabAgent::MeanReward(uint32_t bin, uint32_t actionIndex) const // 计算平均奖励。
{
    const std::deque<double>& buffer = m_rewardBuffers[bin][actionIndex - 1]; // 获取奖励窗口。
    if (buffer.empty()) // 如果没有样本。
    {
        return 0.0; // 返回 0。
    }
    const double sum = std::accumulate(buffer.begin(), buffer.end(), 0.0); // 计算奖励和。
    return sum / static_cast<double>(buffer.size()); // 返回平均值。
}

} // namespace myproject

File: my-project/mab_agent.py
from collections import defaultdict, deque
import random


class MabAgent:
    """按 SNR 分桶的 epsilon-greedy 多臂老虎机。

    这里没有训练神经网络，也不预测未来状态。它把相近 SNR 看作同一种上下文，
    每个 SNR 桶里分别维护每个 action 的最近 reward 均值，然后在“探索”和
    “选择当前均值最高 action”之间切换。
    """

    def __init__(self, num_actions=8, epsilon=0.1, window_size=30):
        self.num_actions = num_actions
        # epsilon 是探索概率：每次决策有 epsilon 的概率随机选一个速率档位。
        self.epsilon = epsilon
        # 每个 action 只保留最近 window_size 个 reward，让算法能跟上移动场景的信道变化。
        self.window_size = window_size
        # SNR 桶边界：(-inf,-5], (-5,0], ... , (35,inf)。
        self.snr_edges = [edge for edge in range(-5, 40, 5)]
        # rewards[snr_bucket][action-1] 是一个 deque，保存该 SNR 桶下该 action 的近期收益。
        self.rewards = defaultdict(lambda: [deque(maxlen=self.window_size) for _ in range(self.num_actions)])

    def _bin(self, snr_db):
        """把连续 SNR 值映射到离散桶，减少学习表规模。"""
        for index, edge in enumerate(self.snr_edges):
            if snr_db <= edge:
                return index
        return len(self.snr_edges)

    def select_action(self, env):
        """根据当前 SNR 桶选择下一次发送速率。

        action 使用 1-based 编号，和 C++ 共享内存结构保持一致。
        """
        self.num_actions = max(1, int(env.numActions))
        bin_id = self._bin(float(env.snrDb))
        buffers = self.rewards[bin_id]

        # 先保证每个 action 至少试过一次，否则均值比较会永久忽略未测试档位。
        untested = [action for action in range(1, self.num_actions + 1) if not buffers[action - 1]]
        if untested:
            return random.choice(untested), self.epsilon

        # 探索：随机尝试一个 action，防止早期偶然 reward 导致算法锁死在次优速率。
        if random.random() < self.epsilon:
            return random.randint(1, self.num_actions), self.epsilon

        # 利用：选择当前 SNR 桶下近期平均 reward 最高的 action。
        means = [
            sum(buffers[action - 1]) / len(buffers[action - 1])
            for action in range(1, self.num_actions + 1)
        ]
        return int(max(range(1, self.num_actions + 1), key=lambda action: means[action - 1])), self.epsilon

    def observe(self, env):
        """把 C++ 上一轮计算出的 reward 写回表格。

        env.lastAction 是上一轮真正发送时用的 action；env.reward 是这个 action
        在当前链路反馈下得到的收益。
        """
        action = int(max(1, min(int(env.lastAction), max(1, int(env.numActions)))))
        self.rewards[self._bin(float(env.snrDb))][action - 1].append(float(env.reward))
